File: Cpp/59.cpp
// Given a positive integer n, generate a square matrix filled with elements
// from 1 to n2 in spiral order.
//
// Example:
//
// Input: 3
// Output:
// [
//  [ 1, 2, 3 ],
//  [ 8, 9, 4 ],
//  [ 7, 6, 5 ]
// ]

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n, vector<int>(n, 0));
        int left = 0, right = n - 1;
        int top = 0, bottom = n - 1;
        int val = 1;

        while (left <= right && top <= bottom) {
            for (int i = left; i <= right && top <= bottom; ++i) {
                res[top][i] = val++;
            }
            top++;

            for (int i = top; i <= bottom && left <= right; ++i) {
                res[i][right] = val++;
            }
            right--;

            for (int i = right; i >= left && top <= bottom; --i) {
                res[bottom][i] = val++;
            }
            bottom--;

            for (int i = bottom; i >= top && left <= right; --i) {
                res[i][left] = val++;
            }
            left++;
        }

        return res;
    }
};

File: Python/59.py
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        left, right, top, bottom = 0, n-1, 0, n-1
        m = [[0]*n for i in range(n)]
        val = 1
        while left<=right and top <= bottom:
            for i in range(left, right+1):
                m[top][i] = val
                val +=1
            top +=1
            for i in range(top, bottom+1):
                m[i][right] = val
                val +=1
            right -=1
            for i in range(right, left-1, -1):
                m[bottom][i] = val
                val +=1
            bottom -=1
            for i in range(bottom, top-1, -1):
                m[i][left] = val
                val +=1
            left +=1
        return m
